### ISSUE TITLE
Improve no-misleading-array-reverse (S4043)

### ISSUE THREAD

**Issue by vilchik-elena** [SonarSource/SonarTS#143]
Rule is added by #80 (for [RSPEC-4043](https://jira.sonarsource.com/browse/RSPEC-4043))

False-Negative
```typescript
const a = [1, 2, 3];
const x = a.reverse().join(); // FN, while "const x = a.reverse()" will raise an issue
```

**Comment by vilchik-elena** [comment]
method `isForbiddenOperation` could be implemented this way:
```
 private isForbiddenOperation(node: ts.Node): boolean {
    return !!node.parent && !is(node.parent, ts.SyntaxKind.ExpressionStatement, ts.SyntaxKind.ReturnStatement);
}
```

But this way we will have FP here:
```
return keys.reverse().join();
```

If `keys` is local variable, we don't care what happens to it inside return statement (so it's FP). On other side, if it's not local, calling `reverse` will change it, and issue is valid in this case.

**Comment by MhdTlb** [comment]
I would like to work on this but I need some guidance though

**Comment by inverno** [comment]
@MhdTlb Great!

You can start by updating this rule's unit test to cover the case where we have a False-Negative :
* Modify https://github.com/SonarSource/SonarTS/blob/master/sonarts-core/tests/rules/noMisleadingArrayReverseRule/noMisleadingArrayReverseRule.lint.ts by adding the False-Negative that @vilchik-elena has described on top of this issue :
``` typescript
function reverseAndJoin() {
  const a = [1, 2, 3];
  const x = a.reverse().join();
  //        ^^^^^^^^^^^ {{Move this array "reverse" operation to a separate statement.}}
}
```
* Run the tests (`yarn test`) to see that now `noMisleadingArrayReverseRule.test.ts` is red
* Locate the method `isForbiddenOperation` in `noMisleadingArrayReverseRule.ts` : https://github.com/SonarSource/SonarTS/blob/master/sonarts-core/src/rules/noMisleadingArrayReverseRule.ts#L104:L115
You see how you could add an extra case to that list of `||` calls? Add a new method call there, something like `isForbiddenChainedCall(node)` and then try to implement it by taking inspiration from the other methods, specifically, consider that `node` will be pointing to `a.reverse()` and its parent should be `a.reverse().join()`.

If you have any questions just drop a line here as usual :) Thanks in advance!

**Comment by MhdTlb** [comment]
@inverno  Thanks for your help. I did push some code I hope it works :) .
I'm using Visual Studio Code and I can't find a way to debug code. Am I missing some settings steps? 
I'm curious what IDE you are using so I can use it too

**Comment by inverno** [comment]
@MhdTlb yes, your code works for the specific case, but it was not yet accounting for @vilchik-elena remark about creating FPs in case of `return array.reverse().join()` (see her second comment in this issue). I started from your commit and added the return case to the unit test. Then I proceeded to fix that FP and to simplify the rule by implementing a small variant of @vilchik-elena idea. You can find your original commit and my addition in #307. As soon as #307 is green I'll merge it and close your initial PR #306.

We use VSC as well and for debugging we typically run jest in watch mode (`yarn test --watchAll`) and play with `console.log`.

Speaking of VSC, the combination of your code and my modification finds a case of useless array deconstruction and reconstruction in Visual Studio Code codebase : https://github.com/Microsoft/vscode/blob/363be8073af349606dbd8a110e720fe2a5aad37d/src/vs/base/common/async.ts#L405
You could consider opening a PR on Visual Studio Code now :)
